File: string_search/rabin_karp/src/RabinKarpDemo.go
package main

import (
	"fmt"
)

// search はテキスト内でパターンを検索し、一致する位置のインデックスのスライスを返します
func search(text, pattern string) []int {
	n := len(text)
	m := len(pattern)
	
	// パターンが文字列より長い場合は検索できない
	if m > n {
		return []int{}
	}
	
	// ハッシュ計算のための定数
	// 大きな素数を使用
	q := 101
	d := 256 // 文字セットのサイズ（ASCIIを想定）
	
	// パターンとテキスト最初のm文字のハッシュ値を計算
	patternHash := 0
	textHash := 0
	h := 1
	
	// h = d^(m-1) mod q の計算
	for i := 0; i < m-1; i++ {
		h = (h * d) % q
	}
	
	// パターンと初期ウィンドウのハッシュ値を計算
	for i := 0; i < m; i++ {
		patternHash = (d*patternHash + int(text[i])) % q
		textHash = (d*textHash + int(text[i])) % q
	}
	
	result := []int{}
	
	// テキスト内を順に探索
	for i := 0; i <= n-m; i++ {
		// ハッシュ値が一致した場合、文字ごとに比較して確認
		if patternHash == textHash {
			// 文字ごとのチェック
			match := true
			for j := 0; j < m; j++ {
				if text[i+j] != pattern[j] {
					match = false
					break
				}
			}
			
			if match {
				result = append(result, i)
			}
		}
		
		// 次のウィンドウのハッシュ値を計算
		if i < n-m {
			// 先頭の文字を削除
			textHash = (d*(textHash-int(text[i])*h) + int(text[i+m])) % q
			
			// 負の値になる場合は調整
			if textHash < 0 {
				textHash += q
			}
		}
	}
	
	return result
}

func main() {
	fmt.Println("RabinKarp TEST -----> start")

	fmt.Println("\nsearch")
	input := []string{"ABABCABCABABABD", "ABABD"}
	fmt.Printf("  入力値: %v\n", input)
	output := search(input[0], input[1])
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsearch")
	input = []string{"AAAAAA", "AA"}
	fmt.Printf("  入力値: %v\n", input)
	output = search(input[0], input[1])
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsearch")
	input = []string{"ABCDEFG", "XYZ"}
	fmt.Printf("  入力値: %v\n", input)
	output = search(input[0], input[1])
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsearch")
	input = []string{"ABCABC", "ABC"}
	fmt.Printf("  入力値: %v\n", input)
	output = search(input[0], input[1])
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nsearch")
	input = []string{"ABC", ""}
	fmt.Printf("  入力値: %v\n", input)
	output = search(input[0], input[1])
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nRabinKarp TEST <----- end")
}

File: string_search/rabin_karp/src/RabinKarpDemo.rb
# Ruby
# 文字列の検索: ラビンカープ法 (Rabin Karp)

def search(text, pattern)
  n = text.length
  m = pattern.length
  
  # パターンが文字列より長い場合は検索できない
  return [] if m > n
  return (0..n-m).to_a if m == 0
  
  # ハッシュ計算のための定数
  # 大きな素数を使用
  q = 101
  d = 256  # 文字セットのサイズ（ASCIIを想定）
  
  # パターンとテキスト最初のm文字のハッシュ値を計算
  pattern_hash = 0
  text_hash = 0
  h = 1
  
  # h = d^(m-1) mod q の計算
  (m-1).times do
    h = (h * d) % q
  end
  
  # パターンと初期ウィンドウのハッシュ値を計算
  m.times do |i|
    pattern_hash = (d * pattern_hash + pattern[i].ord) % q
    text_hash = (d * text_hash + text[i].ord) % q
  end
  
  result = []
  
  # テキスト内を順に探索
  (n - m + 1).times do |i|
    # ハッシュ値が一致した場合、文字ごとに比較して確認
    if pattern_hash == text_hash
      # 文字ごとのチェック
      match = true
      m.times do |j|
        if text[i+j] != pattern[j]
          match = false
          break
        end
      end
      
      result << i if match
    end
    
    # 次のウィンドウのハッシュ値を計算
    if i < n - m
      # 先頭の文字を削除
      text_hash = (d * (text_hash - text[i].ord * h) + text[i + m].ord) % q
      
      # 負の値になる場合は調整
      text_hash += q if text_hash < 0
    end
  end
  
  result
end

def main
  puts "RabinKarp TEST -----> start"

  puts "\nsearch"
  input = ["ABABCABCABABABD", "ABABD"]
  puts "  入力値: #{input}"
  output = search(*input)
  puts "  出力値: #{output}"

  puts "\nsearch"
  input = ["AAAAAA", "AA"]
  puts "  入力値: #{input}"
  output = search(*input)
  puts "  出力値: #{output}"

  puts "\nsearch"
  input = ["ABCDEFG", "XYZ"]
  puts "  入力値: #{input}"
  output = search(*input)
  puts "  出力値: #{output}"

  puts "\nsearch"
  input = ["ABCABC", "ABC"]
  puts "  入力値: #{input}"
  output = search(*input)
  puts "  出力値: #{output}"

  puts "\nsearch"
  input = ["ABC", ""]
  puts "  入力値: #{input}"
  output = search(*input)
  puts "  出力値: #{output}"

  puts "\nRabinKarp TEST <----- end"
end

if __FILE__ == $PROGRAM_NAME
  main
end